llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -3757,11 +3757,21 @@ static unsigned ComputeNumSignBitsImpl(const Value *V,
     }
     case Instruction::Shl: {
       const APInt *ShAmt;
+      Value *X = nullptr;
       if (match(U->getOperand(1), m_APInt(ShAmt))) {
         // shl destroys sign bits.
-        Tmp = ComputeNumSignBits(U->getOperand(0), Depth + 1, Q);
-        if (ShAmt->uge(TyBits) ||   // Bad shift.
-            ShAmt->uge(Tmp)) break; // Shifted all sign bits out.
+        if (ShAmt->uge(TyBits))
+          break; // Bad shift.
+        // We can look through a zext (more or less treating it as a sext) if
+        // all extended bits are shifted out.
+        if (match(U->getOperand(0), m_ZExt(m_Value(X))) &&
+            ShAmt->uge(TyBits - X->getType()->getScalarSizeInBits())) {
+          Tmp = ComputeNumSignBits(X, Depth + 1, Q);
```

Comment:
Regression: https://alive2.llvm.org/ce/z/tcsmgq
Now `src` cannot be folded into `tgt`.
```
define i32 @src(i32 %add.i.i370) {
   %shr.i.i = ashr i32 %add.i.i370, 5
   %conv.i.i371 = zext i32 %shr.i.i to i64
   %shl.i.i = shl i64 %conv.i.i371, 34
    %conv6.i = ashr exact i64 %shl.i.i, 32
   %conv.i41.i = trunc nsw i64 %conv6.i to i32
   ret i32 %conv.i41.i
}

define i32 @tgt(i32 %add.i.i370) {
  %121 = ashr i32 %add.i.i370, 3
   %122 = and i32 %121, -4
   ret i32 %122
}
```

Reply:
I think it is related to this piece of code in InstCombinerImpl::SimplifyDemandedUseBits:

```
  case Instruction::AShr: {
    unsigned SignBits = ComputeNumSignBits(I->getOperand(0), Depth + 1, Q.CxtI);
    ...
      // If the input sign bit is known to be zero, or if none of the top bits
      // are demanded, turn this into an unsigned shift right.
      assert(BitWidth > ShiftAmt && "Shift amount not saturated?");
      APInt HighBits(APInt::getHighBitsSet(
                         BitWidth, std::min(SignBits + ShiftAmt - 1, BitWidth)));
      if (Known.Zero[BitWidth-ShiftAmt-1] ||
          !DemandedMask.intersects(HighBits)) {
        BinaryOperator *LShr = BinaryOperator::CreateLShr(I->getOperand(0),
                                                          I->getOperand(1));
        LShr->setIsExact(cast<BinaryOperator>(I)->isExact());
        LShr->takeName(I);
        return InsertNewInstWith(LShr, I->getIterator());
      }
```

And I do not fully undertstand the logic for the HigBits computation.
Given` %conv6.i = ashr exact i64 %shl.i.i, 32` , I think that BitWidth is 64. And ShifAmt is 32.
So the code is doing
`   APInt::getHighBitsSet(64, std::min(SignBits + 32 - 1, 64)`
And if we know nothing about sign bits we will set the 32 high bits. But the more we know about sign bits, the more bits will be set. With this patch we SignBits==4 so we get more bits set in HighBits.

But then the check `!DemandedMask.intersects(HighBits)` is less likely to be fulfilled. Because the intersects would be more likely to happen if we are able to compute more sign bits. So the better ComputeNumSignBits analysis we have, the less likely that we apply the optimization converting the AShr to LShr. That doesn't make sense to me!

Reply:
Seems to originate from this patch back in 2017: https://github.com/llvm/llvm-project/commit/22178dd33b3460207b8b1dd109339524e50feab5
So the idea actually was to avoid ASHR->LSHR rewrites in certain situations (based on number of known sign bits).

Unfortunately there are no good motivating test cases. Even the test case added in that commit would still pass if reverting the change in SimplifyDemandedUseBits.

Reply:
Created https://github.com/llvm/llvm-project/pull/99155 to avoid such regressions.

Reply:
Rebased branch on top of https://github.com/llvm/llvm-project/commit/b8c4c58ecf186dd91f40bdff4d1bdad403435789 to show that the regression no longer can be seen.